### PR TITLE
Eliminate redundant work in type inference

### LINKED ISSUE #15443: TS v2.3.1 runs out of memory compiling FP library
**TypeScript Version:**: 2.3.1

Repo: https://github.com/gcanti/fp-ts

After upgrading typescript from 2.2.2 to 2.3.1 the compilation (`npm run build`) fails with the following error:

```
<--- Last few GCs --->

   25725 ms: Scavenge 1407.7 (1457.4) -> 1407.7 (1457.4) MB, 46.2 / 0 ms (+ 24.7 ms in 1 steps since last GC) [allocation failure] [incremental marking delaying mark-sweep].
   27079 ms: Mark-sweep 1407.7 (1457.4) -> 1406.4 (1456.4) MB, 1354.1 / 0 ms (+ 1071.3 ms in 617 steps since start of marking, biggest step 25.4 ms) [last resort gc].
   28376 ms: Mark-sweep 1406.4 (1456.4) -> 1405.9 (1457.4) MB, 1297.0 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x1985f0ce3ac1 <JS Object>
    1: recursiveTypeRelatedTo(aka recursiveTypeRelatedTo) [/Users/giulio/.nvm/versions/node/v5.5.0/lib/node_modules/typescript/lib/tsc.js:~27814] [pc=0xf9f8ea8a65b] (this=0x1985f0c04189 <undefined>,source=0x173e2ea53611 <a Type with map 0x1089a6556cc9>,target=0x95cb168fbb1 <a Type with map 0x1089a6556b69>,reportErrors=0x1985f0c04299 <false>)
    2: isRelatedTo(aka isRelatedTo) [/Users/giulio/.n...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
Abort trap: 6
```

Somehow seems related to the modules `Tuple.ts` and `Apply.ts`

In the module `Tuple.ts` if I comment the following lines

```ts
declare module './HKT' {
  // interface HKT<A> {
  //   Tuple: Tuple<any, A>
  // }
  interface HKT2<A, B> {
    Tuple: Tuple<A, B>
  }
}
```

the compilation doesn't crash. 

Alternatively if I don't touch `Tuples.ts` but I comment out all the `liftA*` functions in `Apply.ts` the compilation doesn't crash.

**Expected behavior:**
Compilation success

**Actual behavior:**
Compilation fail

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we eliminate redundant work in type inference. Previously, we would explore instantiations of the same generic type up to five levels deep in type inference (which mirrored what we do in relationship checking to detect infinitely recursive types). We now explore just one level as no further inferences will be made beyond that anyway.

This PR doesn't include a regression test as the original issue only reproduces with the fp-ts library. However, I have manually verified that the out of memory issue is fixed and the compile time for the repro is less than 1s. Also, I have verified there are no changes in the real world code suites.

Fixes #15443.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -10096,21 +10096,29 @@ namespace ts {
...
-            function isInProcess(source: Type, target: Type) {
-                for (let i = 0; i < depth; i++) {
-                    if (source === sourceStack[i] && target === targetStack[i]) {
-                        return true;
+            function isInstantiationInProcess(type: Type) {
+                const symbol = getSymbolForInference(type);
+                if (symbol) {
+                    for (let i = 0; i < depth; i++) {
+                        const t = stack[i];
+                        if (getSymbolForInference(t) === symbol) {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: why not storing `getSymbolForInference(target)` on the stack directly instead of calling it on every entry when we check

**Comment by sledorze** [comment]
will it be part of 2.3.3 ?

**Comment by mhegazy** [comment]
> will it be part of 2.3.3 ?

no. this is TS 2.4 bound.

**Comment by OliverJAsh** [comment]
I'm using Version 2.4.0-dev.20170519 and I've found that the project I shared above no longer has this issue, however I have a separate project that doesn't compile due to this memory issue.

Has anyone else found the same? Will require some work to share a test case.

**Comment by mhegazy** [comment]
Was the other project compiling on 2.3?

**Comment by OliverJAsh** [comment]
@mhegazy Here is the other project that runs out of memory when compiling on 2.4.0-dev.20170519 *and* 2.3.

https://github.com/OliverJAsh/ts-2.4-memory-issue/compare/2.4.0-dev.20170519

**Pull request merged**